File: archives/C#/0144. Binary Tree Preorder Traversal.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public IList<int> PreorderTraversal(TreeNode root) {
        IList<int> result=new List<int>();
        if(root==null)
            return result;
        Stack<TreeNode> rootStack=new Stack<TreeNode>();
        while(root!=null){
            result.Add(root.val);
            if(root.right!=null)
                rootStack.Push(root.right);
            if(root.left!=null)
                root=root.left;
            else if(root.left==null && rootStack.Count()!=0)
                root=rootStack.Pop();
            else
                break;
        }
        return result;
    }
}

File: archives/leetcode2/0144. Binary Tree Preorder Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if not root:
            return []
        
        root_stack=[root]
        rep=[]
        while root_stack:
            cur_root=root_stack.pop()
            rep.append(cur_root.val)
            if cur_root.right:
                root_stack.append(cur_root.right)
            if cur_root.left:
                root_stack.append(cur_root.left)
        return rep
